pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go

```diff
@@ -282,6 +293,121 @@ func configMapDeleted(obj interface{}) {
 	}
 }
 
+// initVolumeHandleToPvcMap performs all the operations required to initialize the volume id to PVC name map.
+// It also watches for PV update & delete operations, and updates the map accordingly.
+func initVolumeHandleToPvcMap(ctx context.Context, k8sClient clientset.Interface) {
+	log := logger.GetLogger(ctx)
+	log.Debugf("Initializing volume ID to PVC name map")
+	k8sOrchestratorInstance.volumeIDToPvcMap = make(map[string]string)
+
+	// Set up kubernetes resource listener to listen events on PersistentVolume
+	k8sOrchestratorInstance.informerManager = k8s.NewInformer(k8sClient)
+	k8sOrchestratorInstance.informerManager.AddPVListener(
+		func(obj interface{}) { // Add
+			pvAdded(obj)
+		},
+		func(oldObj interface{}, newObj interface{}) { // Update
+			pvUpdated(oldObj, newObj)
+		},
+		func(obj interface{}) { // Delete
+			pvDeleted(obj)
+		})
+
+	k8sOrchestratorInstance.informerManager.AddPVCListener(
+		func(obj interface{}) { // Add
+			pvcAdded(obj)
+		},
+		nil, // Update
+		nil, // Delete
+	)
+	k8sOrchestratorInstance.informerManager.Listen()
```

Comment:
Shouldn't we need to make sure that the maps are fully initialized with the current PV/PVC info before returning from this init method? Basically, is there a case where the map is not fully initialized and CSI gets ControllerPublihVolume callbacks? If yes, there is a time window where the ControllerPublishVolume may not fake attach since the map is not populated yet.

Reply:
Doesn't `Listen()` method in informerManager takes care of that? It waits for the cache to be synced before returning.
```
if im.pvSynced != nil && im.pvcSynced != nil && im.podSynced != nil && im.configMapSynced != nil {
      if !cache.WaitForCacheSync(im.stopCh, im.pvSynced, im.pvcSynced, im.podSynced, im.configMapSynced) {
	  return
      }
}
```

Reply:
Can you point me to the code where this is handled?

Reply:
https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/0f0e83f88be00c0baa85537f11ff8bd3cbdf3ace/pkg/kubernetes/informers.go#L154